### DISCUSSION TITLE
Surface Go 4 on Debian

### DISCUSSION THREAD

**Discussion by debsurf** [linux-surface/linux-surface#1269] [Q&A]
Today my Surface Go 4 arrived. 256 GB with Win10P.

A live-stick with Debian 12 (Gnome) was a big success: Almost everything works right out of the box. Suspend/Resume with the Type Cover closed/open works perfectly, also the change between portrait and landscape mode, battery status is shown correct, my i-tec metal nano dock (3x USB 3.0, HDMI, SD/TF-Card, Thunderbolt, Audio, RJ-45) also works without problems.

After disabling "secure core" in the BIOS, it also boots directly from USB.

Not working so far with the stock kernel: Buttons and Cameras (no surprise).

In Win10, I installed all the updates first (which took about 2 hours...), then I shrinked the Windows-Partition so there is enough space for Linux. I started Bitlocker, then disabled it and waited until decryption finished.

Problem now: The debian installer does not recognize the UFS-Drive. When I get to the partitioning, all I see is a 8 GB SanDisk Cruzer which is my Installer-Stick. I tried the expert install, because it offers some additional modules to load. But I found nothing suitable there. Any hints? The BIOS has no Options for the drive.

The Ubuntu-Installer recognizes the UFS-Drive: My Installer-Stick is sda, the internal drive appears as sdb and I can see the Windows-Partitions and the free space there. But I really would like to install a debian...

**Comment by debsurf** [comment]
Update: I saw that the calamares-Installer on the debian live-system detects the internal UFS-drive correctly. So I was able to install debian stable bookworm 12.2.

Now the power button works, the tablet goes into suspend or wakes up when I press it.

As expected, still not working: The volume up/down-buttons and the cameras. But I don't need them, so I will not include the patched surface-kernel. (All the buttons on the type cover work as expected)

Of course Firefox needs extra help with setting MOZ_ENABLE_WAYLAND=1 in a file under /etc/environment.d/ but then touchgestures work perfectly.

In short: Thanks to all the folks who put time and effort in this for the last years. Even when the Go 4 is still brandnew, the linux experience is really good. I will post some hardware details and battery testing results in the next days...

**Comment by debsurf** [comment]
I noticed a strange behavoir: After logging in, Gnome always was turned to the right (portrait) mode, even when I had the tablet in landscape mode with attached type cover. So after every boot or resume, I hat to turn the hole thing left, wait a second and turn it right again.

So I upgraded to debian testing with gnome 44 (transition to 45 ongoing). After reboot, the GUI did not came up, i got just a console. Entry in dmesg:

ucsi_acpi USBC000:00: error -ETIMEDOUT: PPM init failed

So I decided to add the surface-repo. Problem: The key was not accepted by apt, so I had to install the surface-packages with apt-get and the --allow-unauthenticated flag.
Then I installed the secureboot-mok.

After next boot...nothing.
He writes just, that he cannot find any /efi or /mok, and "something has gone seriously wrong". I can still get into BIOS, but I am not able anymore  to boot any USB-Stick. Even if I turn secure boot off, it does not boot....

**Comment by debsurf** [comment]
I can still boot windows 10, even boot from a USB-Stick with Rescue Windows, but cannot boot anymore from any Linux-Installer-Stick. So I wiped the hole efi-partition in Win10 and made a new one with windows-tools, still no difference. When I insert a Linux-Stick, it always shows this message for a few seconds:
![uefi](https://github.com/linux-surface/linux-surface/assets/147057336/6bbbc36e-17fe-4caf-ab31-2bb63081a760)

Failed to open \EFI\BOOT\amd64.efi - Not found
Failed to load image : Not Found
Failed to start MokManager: Not Found
Something has gone seriously wrong: Import_mok_state() failed: Not Found

Anyone maybe knows a way to rescue this tablet?

**Reply by qzed** [reply]
That looks like something is wrong with your EFI setup. In particular, it looks like you're booting into shim, but shim can't verify the signature of the follow-up EFI executable and fails to find MokManager. I think reinstalling linux / preparing a new linux stick from scratch is the easiest option here.

**Reply by StollD** [reply]
This is what happens if you enroll a key and then don't confirm the enrollment but instead boot from a USB drive.

The shim on the USB drive will see the enrollment request and try to launch MokManager. But Live ISOs don't usually ship with MokManager, so shim fails to launch it.

AFAIK the enrollment has a timeout so at some point it should just work.

However, I believe https://github.com/linux-surface/linux-surface/issues/1162 also plays a role here. I think that MokManager has a similar issue as the Fedora grub and it just won't launch if the UEFI firmware is in NX mode.

**Reply by debsurf** [reply]
Ok, I will try tonight if the stick still boots after some waiting time.

**Reply by qzed** [reply]
How is the MOK enrollment triggered? Is there some UEFI variable that one could reset?

**Comment by debsurf** [comment]
I would love to, but I can't. I get the mentioned message when I try to boot from the debian installer-stick, then the tablet turns off. If I try to boot from any other linux-stick (ubuntu, live-debian, fedora), I just see the windows-logo and the system freezes.

Still possible: Boot into win10. So I guess I need a repair option which works from windows. The efi-partition is clean, but somehow I need to clean the internal UEFI-settings. I can still get into BIOS, but can't see an appropriate option there to do that...

**Comment by debsurf** [comment]
Ok: I tried the Microsoft Surface Data Eraser, which I hoped would maybe also set the UEFI to factory reset. Unfortunately, it doesn't. It just swiped the disk. I installed Win10P from scratch and let him do all the firmware updates, again in hope he might reset UEFI-Settings. Now I have a perfekt Win10-install, but everything else is still the same:

Inserting a debian-installstick oder debian-live, he prints the mentioned message (failed to open \EFI\BOOT...) and turns power off after a few second.
Inserting any other linux-stick, he hangs with windows logo. Let him hang during the night with ubuntu-stick in hope there is a time out, but 8  hours later there is still only a beautiful windows logo.

I can still go into BIOS, boot Windows or boot from Windows-Stick.

Any ideas? Looks like the secureboot-mok-package is harmful, maybe put a warning to it not to use it on a Go 4...

**Reply by StollD** [reply]
You could try ventoy, it doesn't ship with shim at all, so it wont be affected. But you need to disable secureboot.

Once you are booted into a linux system you can try running `sudo mokutil --revoke-import`.

And yeah we should probably mark secureboot enrollment as broken on newer UEFI firmwares in the wiki.

**Reply by debsurf** [reply]
Great, that worked. In a very strange way.

I put the ventoy iso on a stick, booted from it. It offered me to install (whatever), but did not detect the ufs-drive. Probably it lacks the ufshcd-modules like the debian installer. So i choose a second stick as target to install, after that i booted into that second stick. And then a blue screen appeared, asked me if I want to enroll the mok. Yes...and I entered the surface-password.

And after next reboot, I am back on square one: The tablet now boots all the install-sticks normaly, no more lock. Puh :)

So the injection of mok generally works on a Go 4, for some strange reason not very straightforward...

**Reply by qzed** [reply]
Looks like a similar thing happens on the SLG3: https://github.com/linux-surface/linux-surface/issues/1275#issuecomment-1771980841

**Comment by Pastitas** [comment]
@debsurf  where you able to boot a linux system after installing to the main drive? I cannot seem to be able to (arch and fedora tested)
 
Arch installs fine after disabling secure-boot but gets stuck waiting for the root drive ```gave up waiting for root file system device```

>>I'm wondering if maybe this has to do with the drives and the format, microsoft specs mention UFS drives but afaik that's just the filesystem, not the drive technology. When listed in the installer live media they appear as a normal /dev/sdb device, so I'm assuming it's an ssd but It seems like it could be related

edit: this is wrong, UFS stands for Universal Flash Storage and it's a drive technology, it's just not common in laptops.

**Reply by Pastitas** [reply]
I've been reading up and It seems it might be just necessary to add ufs support to the initramfs so it can mount the drive, It's weird but it makes sense that it's not default since it's a rare use-case. @qzed have you seen this before in any other surface device? Here are some docs:

https://www.kernel.org/doc/html/latest/scsi/ufs.html
https://bbs.archlinux.org/viewtopic.php?id=259176

**Reply by qzed** [reply]
I think something with the comments got mixed up, so here again: I think the Go 4 is the first with UFS. The earlier Go devices had eMMC if I remember correctly. So I haven't seen this before.

So maybe we should build in the UFS driver in our kernels? Then you wouldn't need to manually add it to the initramfs. Although you'd still need to do that for the stock distro kernels.

**Reply by Pastitas** [reply]
It might help but for the distros that I tried except for linux mint debian edition the problem was before first boot, so instructions on installing the linux-surface kernel from a chroot environment from the live-usb might make sense.

**Reply by qzed** [reply]
Feel free to add that to the wiki (maybe on a separate page, to not confuse people in the standard install instructions).

**Reply by Pastitas** [reply]
I'm not sure how that would be done anyway, since I didn't need to do it for linux mint. If I come around to installing arch like I have on my main setup and do this, I'll make a pr to the docs

**Comment by Pastitas** [comment]


**Reply by qzed** [reply]
I think the Go 4 is the first with UFS. The earlier Go devices had eMMC if I remember correctly. So I haven't seen this before.

**Comment by debsurf** [comment]
Debian Installer also lacks ufs-support so far. See bug 1053937 at debian bug tracker...

**Comment by Pastitas** [comment]
Well, that was oddly easy, LMDE 6 (linux mint debian edition) install worked and booted flawlessly, I'll do a test of features but it's perfect so far (keyboard cover and pen work)

**Reply by GreatBigWhiteWorld** [reply]
I tried several distros and can't pass the first boot after installation, with similar errors about not being able to find drives.
You found that Mint Debian edition would work? Or you also made some other distros work? If so can you share the general steps?

**Reply by Pastitas** [reply]
LMDE 6 and nixos installers didn't require any extra steps and are confirmed to work
